internal/states/statefile/version4.go

```diff
@@ -137,6 +144,30 @@ func prepareStateV4(sV4 *stateV4) (*File, tfdiags.Diagnostics) {
 				key = addrs.NoKey
 			}
 
+			if isV4.ProviderConfig != "" && rsV4.ProviderConfig != "" {
+				diags = diags.Append(tfdiags.Sourceless(
+					tfdiags.Error,
+					"Provider field conflict in state",
+					fmt.Sprintf("Resource %s has a provider address, as well as an instance with a provider address.", rAddr.Absolute(moduleAddr)),
+				))
+			}
+
+			if isV4.ProviderConfig == "" && rsV4.ProviderConfig == "" {
+				diags = diags.Append(tfdiags.Sourceless(
+					tfdiags.Error,
+					"Provider field missing state",
+					fmt.Sprintf("Resource %s is missing a provider address, both on the resource and the resource instances.", rAddr.Absolute(moduleAddr)),
+				))
+			}
```

Comment:
This seems a little different to how things settled in the RFC, though we could of course choose to update the RFC to match this code rather than the other way around.

The way this was presented in the RFC was for the _reader_ to be more tolerant with logic something like this:

```go
if isV4.ProviderConfig == "" {
    if rsV4.ProviderConfig == "" {
        diags = diags.Append(tfdiags.Sourceless(
            tfdiags.Error,
            "Invalid resource instance in state",
            fmt.Sprintf(
                "Resource instance %s has no provider instance address, and neither does its containing resource.",
                rAddr.Absolute(moduleAddr).Instance(key)),
            ),
        ))
        continue
    }

    // Older versions of the state snapshot format required all instances
    // to have the same provider instance address and stored it only
    // once at the resource level, so if we don't have an instance-level
    // provider instance address then we'll use the one from the resource
    // level instead.
    isV4.ProviderConfig = rsV4.ProviderConfig
}

// (and then everything below can just assume we're always
// using isV4.ProviderConfig.)
```

I think both ways are reasonable. I wrote it the way I did in the RFC because I was trying to follow the robustness principle: the code that _writes_ state snapshots is intentionally fussy about _either_ populating at the resource level or the instance level, but there isn't really any major downside (as far as I can see) of tolerating mismatches when we _read_ the state, and being flexible gives us the most latitude to make different design tradeoffs later if we learn of new requirements we're not considering today.

Reply:
My main concern with reading state with mismatching keys is that we don't know which one the user actually is trying to select.  It stems from either someone manually editing the state file or a programmatic bug in OpenTofu.

I'd rather have the reader be a bit more picky about this scenario as it can represent a serious issue.

Reply:
Discussed in core team meeting, will change this to a warning

Reply:
Fixed in 8be7bcc48e

Reply:
It seems like we're settled on the "update the RFC to match the implementation" side of this, and so I'll add an item to https://github.com/opentofu/opentofu/issues/2123 to represent that since we can merge this PR before we update the RFC.